File: services/chat-agent/integrations/jira.py
# integrations/jira.py — read-only Jira Cloud adapter.
#
# Auth: HTTP Basic with email + API token.
#   Jira Cloud does not accept password auth; you need an API token from
#   https://id.atlassian.com/manage-profile/security/api-tokens
#
# API: Jira Cloud REST v3 (the current version; v2 is deprecated).
#   Docs: https://developer.atlassian.com/cloud/jira/platform/rest/v3/
#
# Description format: Jira v3 returns the `description` field as Atlassian
# Document Format (ADF) — a JSON tree, not a plain string.  We walk the tree
# and concatenate all `text` leaf nodes.  The result won't preserve rich
# formatting but is good enough for RAG ingestion.
#
# Pagination: /search uses startAt + maxResults.  We page until fewer than
# maxResults issues are returned (last page).

from datetime import datetime

import logging

import httpx

from integrations.base import Item, PMIntegration

logger = logging.getLogger("uvicorn.error")

# How many issues to fetch per API call.  Jira's hard cap is 100.
_PAGE_SIZE = 100


def _extract_adf_text(node: dict) -> str:
    """Recursively extract plain text from an Atlassian Document Format node.

    ADF is a tree where leaves have type="text" and a "text" field.  All
    other node types (paragraph, heading, bulletList, ...) only have a
    "content" list of child nodes.  We walk the whole tree and join the
    leaf text with newlines between block-level nodes.
    """
    if node.get("type") == "text":
        return node.get("text", "")
    children = node.get("content", [])
    parts = [_extract_adf_text(child) for child in children]
    # Drop empty strings so we don't get blank lines from empty containers.
    return "\n".join(p for p in parts if p)


def _jira_date(iso: str) -> str:
    """Convert ISO-8601 (e.g. '2024-01-15T10:30:00Z') to Jira JQL date format.

    Jira's JQL date parser accepts 'YYYY-MM-DD HH:mm'.  We strip seconds and
    timezone info because Jira interprets the date in the project's timezone.
    """
    try:
        dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d %H:%M")
    except ValueError:
        # If parsing fails, pass the string through and let Jira reject it.
        return iso


class JiraIntegration(PMIntegration):
    """Fetch issues from a Jira Cloud project.

    external_ref key: "jira_project_key" (e.g. "ALPHA").
    """

    def __init__(
        self,
        base_url: str,
        email: str,
        api_token: str,
        *,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        # base_url looks like "https://your-org.atlassian.net"
        self._base_url = base_url.rstrip("/")
        # httpx Basic auth tuple: (username, password)
        self._auth = (email, api_token)
        # Injected transport for unit tests; None means real network.
        self._transport = transport

    def _client(self) -> httpx.AsyncClient:
        """Return a configured AsyncClient, reused by all write methods."""
        return httpx.AsyncClient(
            auth=self._auth,
            timeout=30.0,
            transport=self._transport,
        )

    @staticmethod
    def _text_to_adf(text: str) -> dict:
        """Wrap plain text in an Atlassian Document Format doc node.

        Jira v3 requires comment bodies to be ADF objects, not plain strings.
        Each non-empty line becomes its own paragraph node.  This is the
        inverse of _extract_adf_text above.
        """
        lines = [line for line in text.splitlines() if line]
        if not lines:
            lines = [text]
        paragraphs = [
            {
                "type": "paragraph",
                "content": [{"type": "text", "text": line}],
            }
            for line in lines
        ]
        return {"version": 1, "type": "doc", "content": paragraphs}

    async def fetch_items(
        self,
        external_ref: dict,
        updated_since: str | None = None,
    ) -> list[Item]:
        key = external_ref.get("jira_project_key")
        if not key:
            return []

        jql = f"project = {key} ORDER BY updated ASC"
        if updated_since:
            jql = (
                f'project = {key} AND updated >= "{_jira_date(updated_since)}" '
                f"ORDER BY updated ASC"
            )

        items: list[Item] = []
        start_at = 0

        async with self._client() as client:
            # The new /search/jql API uses cursor-based pagination via
            # nextPageToken.  We loop until the response contains no token.
            next_page_token: str | None = None
            while True:
                body: dict = {
                    "jql": jql,
                    "fields": ["summary", "description", "status", "assignee", "updated"],
                    "maxResults": _PAGE_SIZE,
                }
                if next_page_token:
                    body["nextPageToken"] = next_page_token

                resp = await client.post(
                    f"{self._base_url}/rest/api/3/search/jql",
                    json=body,
                )
                resp.raise_for_status()
                data = resp.json()
                issues = data.get("issues", [])

                for issue in issues:
                    fields = issue["fields"]

                    desc = fields.get("description") or ""
                    if isinstance(desc, dict):
                        # Jira v3 ADF object — extract plain text.
                        desc = _extract_adf_text(desc)

                    assignee_node = fields.get("assignee") or {}
                    items.append(
                        Item(
                            id=issue["key"],
                            title=fields.get("summary", ""),
                            body=desc,
                            status=fields.get("status", {}).get("name", ""),
                            assignee=assignee_node.get("displayName"),
                            url=f"{self._base_url}/browse/{issue['key']}",
                            updated_at=fields.get("updated", ""),
                            raw=issue,
                        )
                    )

                next_page_token = data.get("nextPageToken")
                if not next_page_token or len(issues) < _PAGE_SIZE:
                    break

        return items

    async def add_comment(
        self, external_ref: dict, item_id: str, body: str
    ) -> dict:
        """Post a comment on a Jira issue.

        Args:
            external_ref: Must contain "jira_project_key" (used only for context;
                          the actual target is item_id).
            item_id:      The Jira issue key, e.g. "ALPHA-12".
            body:         Plain-text comment.  Wrapped into ADF before sending.

        Returns:
            {id, url, created_at} from the Jira response.
        """
        async with self._client() as client:
            resp = await client.post(
                f"{self._base_url}/rest/api/3/issue/{item_id}/comment",
                json={"body": self._text_to_adf(body)},
            )
            resp.raise_for_status()
            data = resp.json()
        return {
            "id": data["id"],
            # focusedCommentId deep-links the browser directly to the new comment.
            "url": f"{self._base_url}/browse/{item_id}?focusedCommentId={data['id']}",
            "created_at": data.get("created", ""),
        }

File: services/mcp-server/internal/tools/jira.go
package tools

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"

	"mcp-server/internal/mcp"
)

type jiraClient struct {
	baseURL string
	email   string
	token   string
}

func jiraIsConfigured() bool {
	return os.Getenv("JIRA_BASE_URL") != "" &&
		os.Getenv("JIRA_EMAIL") != "" &&
		os.Getenv("JIRA_API_TOKEN") != ""
}

func newJiraClient() *jiraClient {
	return &jiraClient{
		baseURL: strings.TrimRight(os.Getenv("JIRA_BASE_URL"), "/"),
		email:   os.Getenv("JIRA_EMAIL"),
		token:   os.Getenv("JIRA_API_TOKEN"),
	}
}

func (c *jiraClient) get(path string) ([]byte, int, error) {
	req, err := http.NewRequest("GET", c.baseURL+path, nil)
	if err != nil {
		return nil, 0, err
	}
	req.SetBasicAuth(c.email, c.token)
	req.Header.Set("Accept", "application/json")
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(resp.Body, 100*1024))
	return body, resp.StatusCode, nil
}

func (c *jiraClient) post(path string, payload any) ([]byte, int, error) {
	b, _ := json.Marshal(payload)
	req, err := http.NewRequest("POST", c.baseURL+path, strings.NewReader(string(b)))
	if err != nil {
		return nil, 0, err
	}
	req.SetBasicAuth(c.email, c.token)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(resp.Body, 100*1024))
	return body, resp.StatusCode, nil
}

func (c *jiraClient) searchIssues(args map[string]any) (mcp.ToolCallResult, error) {
	query, errResult, err := requireString(args, "query")
	if errResult != nil {
		return *errResult, err
	}
	maxResults := int(optionalFloat(args, "max_results", 20))

	path := fmt.Sprintf("/rest/api/3/issue/search?jql=%s&maxResults=%d&fields=summary,status,assignee",
		url.QueryEscape(query), maxResults)

	raw, status, err := c.get(path)
	if err != nil {
		return textErr(fmt.Sprintf("jira request failed: %v", err))
	}
	if status != 200 {
		return textErr(fmt.Sprintf("jira error %d: %s", status, string(raw)))
	}

	var result struct {
		Total  int `json:"total"`
		Issues []struct {
			Key    string `json:"key"`
			Fields struct {
				Summary  string                                             `json:"summary"`
				Status   struct{ Name string `json:"name"` }               `json:"status"`
				Assignee *struct{ DisplayName string `json:"displayName"` } `json:"assignee"`
			} `json:"fields"`
		} `json:"issues"`
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		return textErr(fmt.Sprintf("parse response failed: %v", err))
	}

	type issueOut struct {
		Key      string `json:"key"`
		Summary  string `json:"summary"`
		Status   string `json:"status"`
		Assignee string `json:"assignee,omitempty"`
		URL      string `json:"url"`
	}
	out := make([]issueOut, len(result.Issues))
	for i, iss := range result.Issues {
		assignee := ""
		if iss.Fields.Assignee != nil {
			assignee = iss.Fields.Assignee.DisplayName
		}
		out[i] = issueOut{
			Key:      iss.Key,
			Summary:  iss.Fields.Summary,
			Status:   iss.Fields.Status.Name,
			Assignee: assignee,
			URL:      c.baseURL + "/browse/" + iss.Key,
		}
	}
	return textResult(map[string]any{"total": result.Total, "issues": out})
}

func (c *jiraClient) getIssue(args map[string]any) (mcp.ToolCallResult, error) {
	key, errResult, err := requireString(args, "key")
	if errResult != nil {
		return *errResult, err
	}

	path := fmt.Sprintf("/rest/api/3/issue/%s?fields=summary,description,status,assignee,priority,created,updated",
		url.PathEscape(key))

	raw, status, err := c.get(path)
	if err != nil {
		return textErr(fmt.Sprintf("jira request failed: %v", err))
	}
	if status != 200 {
		return textErr(fmt.Sprintf("jira error %d: %s", status, string(raw)))
	}

	var result struct {
		Key    string `json:"key"`
		Fields struct {
			Summary     string                                             `json:"summary"`
			Description any                                                `json:"description"`
			Status      struct{ Name string `json:"name"` }               `json:"status"`
			Assignee    *struct{ DisplayName string `json:"displayName"` } `json:"assignee"`
			Priority    *struct{ Name string `json:"name"` }              `json:"priority"`
			Created     string                                             `json:"created"`
			Updated     string                                             `json:"updated"`
		} `json:"fields"`
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		return textErr(fmt.Sprintf("parse response failed: %v", err))
	}

	assignee, priority := "", ""
	if result.Fields.Assignee != nil {
		assignee = result.Fields.Assignee.DisplayName
	}
	if result.Fields.Priority != nil {
		priority = result.Fields.Priority.Name
	}

	return textResult(map[string]any{
		"key":         result.Key,
		"summary":     result.Fields.Summary,
		"description": adfToText(result.Fields.Description),
		"status":      result.Fields.Status.Name,
		"assignee":    assignee,
		"priority":    priority,
		"url":         c.baseURL + "/browse/" + result.Key,
		"created":     result.Fields.Created,
		"updated":     result.Fields.Updated,
	})
}

func (c *jiraClient) addComment(args map[string]any) (mcp.ToolCallResult, error) {
	key, errResult, err := requireString(args, "key")
	if errResult != nil {
		return *errResult, err
	}
	body, errResult, err := requireString(args, "body")
	if errResult != nil {
		return *errResult, err
	}

	// Jira requires plain text wrapped in Atlassian Document Format (ADF).
	payload := map[string]any{
		"body": map[string]any{
			"type":    "doc",
			"version": 1,
			"content": []any{
				map[string]any{
					"type": "paragraph",
					"content": []any{
						map[string]any{"type": "text", "text": body},
					},
				},
			},
		},
	}

	path := fmt.Sprintf("/rest/api/3/issue/%s/comment", url.PathEscape(key))
	raw, status, err := c.post(path, payload)
	if err != nil {
		return textErr(fmt.Sprintf("jira request failed: %v", err))
	}
	if status != 201 {
		return textErr(fmt.Sprintf("jira error %d: %s", status, string(raw)))
	}

	var result struct {
		ID      string `json:"id"`
		Created string `json:"created"`
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		return textErr(fmt.Sprintf("parse response failed: %v", err))
	}

	return textResult(map[string]any{
		"comment_id": result.ID,
		"url":        c.baseURL + "/browse/" + key + "?focusedCommentId=" + result.ID,
		"created_at": result.Created,
	})
}

// adfToText extracts plain text from Atlassian Document Format (ADF).
func adfToText(node any) string {
	if node == nil {
		return ""
	}
	var sb strings.Builder
	extractADFNode(node, &sb)
	return strings.TrimSpace(sb.String())
}

func extractADFNode(node any, sb *strings.Builder) {
	m, ok := node.(map[string]any)
	if !ok {
		return
	}
	if t, _ := m["type"].(string); t == "text" {
		if text, _ := m["text"].(string); text != "" {
			sb.WriteString(text)
		}
		return
	}
	if children, ok := m["content"].([]any); ok {
		for _, child := range children {
			extractADFNode(child, sb)
		}
		sb.WriteString(" ")
	}
}

func jiraDefinitions() []mcp.ToolDefinition {
	return []mcp.ToolDefinition{
		{
			Name:        "jira_search_issues",
			Description: "Search Jira issues using a JQL query. Returns matching issues with their key, summary, status, assignee, and URL.",
			InputSchema: mcp.JSONSchema{
				Type: "object",
				Properties: map[string]mcp.Property{
					"query":       {Type: "string", Description: `JQL query string, e.g. "project = PROJ AND status = 'In Progress' ORDER BY created DESC".`},
					"max_results": {Type: "number", Description: "Maximum number of issues to return (1–50, default 20)."},
				},
				Required: []string{"query"},
			},
		},
		{
			Name:        "jira_get_issue",
			Description: `Fetch full details for a single Jira issue by its key (e.g. "PROJ-123"). Returns summary, description, status, assignee, priority, and timestamps.`,
			InputSchema: mcp.JSONSchema{
				Type: "object",
				Properties: map[string]mcp.Property{
					"key": {Type: "string", Description: `Jira issue key, e.g. "PROJ-123".`},
				},
				Required: []string{"key"},
			},
		},
		{
			Name:        "jira_add_comment",
			Description: "Post a comment on a Jira issue. Returns the new comment ID and URL.",
			InputSchema: mcp.JSONSchema{
				Type: "object",
				Properties: map[string]mcp.Property{
					"key":  {Type: "string", Description: `Jira issue key, e.g. "PROJ-123".`},
					"body": {Type: "string", Description: "Comment text to post."},
				},
				Required: []string{"key", "body"},
			},
		},
	}
}
